src/main/webapp/script.js

```diff
@@ -0,0 +1,232 @@
+    // if image is valid - put it on page
+    const imageUrl = URL.createObjectURL(event.target.files[0]);
+    putImageOnPage(imageUrl);
+  }).catch((error) => {
+    // if image is not valid - display error message
+    alert(error.message);
+  });
+}
+
+/**
+ * Function to validate the uploaded image.
+ * If valid - returns nothing, if not - throws error.
+ * @throws {Error} validation errors
+ */
+async function validateImageUpload() {
+  // get FileList with all the files from input element.
+  const files = document.getElementById('upload-image').files;
+
+  // check if no files were uploaded.
+  if (files.length === 0) {
+    throw new Error('Nothing is uploaded');
+  }
+
+  // get first (and the only) file from FileList.
+  const file = files[0];
+
+  // check whether file is jpeg or png.
+  const fileType = await getImageTypeOrError(file);
+
+  // check file size.
+  validateImageSize(file, fileType);
+
+  // Image resolution can not be more than 1920x1080.
+  await validateImageResolution(file);
+}
+
+/**
+ * Function to make sure the image is not too big.
+ * @param {File} image
+ * @param {string} imageType Can be 'png' or 'jpeg' only.
+ */
+function validateImageSize(image, imageType) {
+  // This limit comes from simple calculation.
+  // MAX_RESOLUTION * AVERAGE_NUMBER_OF_BYTES_PER_PIXEL_IN_PNG
+  // 1920 x 1080 x 4 ~ 8294400 bytes ~ 8 mb
+  const PNG_LIMIT_MB = 8;
+  if (imageType === 'png' && image.size > PNG_LIMIT_MB * 1024 * 1024) {
+    throw new Error(
+        `File size should not exceed ${PNG_LIMIT_MB}MB for png images.` +
+        'The size of an uploaded png image is ' +
+        Math.ceil(image.size / 1024 / 1024) + 'MB');
+  }
+
+  // This limit comes from similar calculation.
+  // MAX_RESOLUTION * AVERAGE_NUMBER_OF_BITS_PER_PIXEL_IN_JPG
+  // 1920 x 1080 x 8.25 ~ 17107200 bits ~ 2 mb
+  const JPG_LIMIT_MB = 2;
+  if (imageType === 'jpeg' && image.size > JPG_LIMIT_MB * 1024 * 1024) {
+    throw new Error(
+        `File size should not exceed ${JPG_LIMIT_MB}MB for jpeg images.` +
+        'The size of an uploaded jpeg image is ' +
+        Math.ceil(image.size / 1024 / 1024) + 'MB');
+  }
+}
+
+/**
+ * Function to get file type if it is png or jpeg
+ * and throw error otherwise
+ * @param {File} file
+ * @return {Promise}
+ */
+function getImageTypeOrError(file) {
+  return new Promise(function(resolve, reject) {
+    const fileReader = new FileReader();
+
+    fileReader.onloadend = function(e) {
+      // get first 4 bytes from file - they contain information
+      // about the extension.
+      const bytes = new Uint8Array(e.target.result);
+
+      // convert those bytes to string
+      let header = '';
+      for (let i = 0; i < bytes.length; i++) {
+        header += bytes[i].toString(16);
+      }
+
+      const PNG_HEADERS = ['89504e47'];
+      if (PNG_HEADERS.includes(header)) {
+        resolve('png');
+      }
+
+      const JPEG_HEADERS = ['ffd8ffe0', 'ffd8ffe1',
```

Comment:
I think it would be a good coding practice to declare these const variables, either in a single file or at the top of the file?

What do you think Mikhail?

Reply:
I think they are good in here as they are referenced only once and defined within the relevant context.